remotespark/sparkkernelbase.py

```diff
@@ -28,46 +35,18 @@ def __init__(self, **kwargs):
         super(SparkKernelBase, self).__init__(**kwargs)
         self.logger = Log(self.client_name)
         self.already_ran_once = False
+        self._fatal_error = None
 
         # Disable warnings for test env in HDI
         requests.packages.urllib3.disable_warnings()
 
-    def get_configuration(self):
-        try:
-            username = get_configuration(self.username_conf_name)
-            password = get_configuration(self.password_conf_name)
-            url = get_configuration(self.url_conf_name)
-            return username, password, url
-        except KeyError:
-            error = "FATAL ERROR: Please set configuration for '{}', '{}', '{} to initialize Kernel.".format(
-                self.username_conf_name, self.password_conf_name, self.url_conf_name)
-            stream_content = {"name": "stdout", "text": error}
-            self.send_response(self.iopub_socket, "stream", stream_content)
-            self.logger.error(error)
-
-            self.do_shutdown(False)
-
-            # Not reachable because shutdown will kill the kernel
-            raise
-
-    def initialize_magics(self, username, password, url):
-        connection_string = get_connection_string(url, username, password)
-
-        register_magics_code = "%load_ext remotespark"
-        self.execute_cell_for_user(register_magics_code, True, False)
-        self.logger.debug("Loaded magics.")
-
-        add_endpoint_code = "%spark add {} {} {} skip".format(
-            self.client_name, self.session_language, connection_string)
-        self.execute_cell_for_user(add_endpoint_code, True, False)
-        self.logger.debug("Added endpoint.")
-
-        self.already_ran_once = True
-
     def do_execute(self, code, silent, store_history=True, user_expressions=None, allow_stdin=False):
+        if self._fatal_error is not None:
+            self._abort_with_fatal_error(self._fatal_error)
+
         if not self.already_ran_once:
-            (username, password, url) = self.get_configuration()
-            self.initialize_magics(username, password, url)
+            (username, password, url) = self._get_configuration()
```

Comment:
Idea: the _get_configuration method is only called once, here, and its output is passed directly to the _initialize_magics method. So maybe _initialize_magics should take no arguments and call _get_configuration itself.